File: java/veci.java
import java.util.Scanner;

public class veci {
	static int permuted = Integer.MAX_VALUE;

	static void getPer(char[] X, int k, int m, int number) {
		if (k == m) {
			int res = Integer.parseInt(new String(X));
			if (res > number && res < permuted)
				permuted = res;
		} else
			for (int i = k; i <= m; i++) {
				swap(X, k, i);
				getPer(X, k + 1, m, number);
				swap(X, k, i);
			}
	}

	static void swap(char[] x, int k, int i) {
		char tmp = x[k];
		x[k] = x[i];
		x[i] = tmp;
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int X = scan.nextInt();
		char[] X_array = String.valueOf(X).toCharArray();
		getPer(X_array, 0, X_array.length - 1, X);
		System.out.println(permuted == Integer.MAX_VALUE ? "0" : permuted);
		scan.close();
	}
}

File: ts/src/veci.ts
import readline = require('readline')

function getResult (nums: number[]): number {
  const original = parseInt(nums.join(''))
  getPer(nums, 0, original)
  return permuted === Number.MAX_SAFE_INTEGER ? 0 : permuted
}

function getPer (nums: number[], start: number, num: number) {
  if (start === nums.length - 1) {
    const res = parseInt(nums.join(''))
    if (res > num) {
      permuted = Math.min(permuted, res)
    }
  } else {
    for (let i = start; i < nums.length; i++) {
      nums = swap(nums, start, i)
      getPer(nums, start + 1, num)
      nums = swap(nums, start, i)
    }
  }
}

function swap (nums: number[], i: number, j: number): number[] {
  const temp = nums.slice()
  ;[temp[i], temp[j]] = [temp[j] ?? 0, temp[i] ?? 0]
  return temp
}

let permuted = Number.MAX_SAFE_INTEGER

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    console.log(getResult(line.split('').map(Number)))
  })
  .on('close', () => {
    process.exit(0)
  })
